qiime2/sdk/plugin_manager.py

```diff
@@ -62,22 +67,39 @@ def _init(self):
         for plugin in self.plugins.values():
             self._integrate_plugin(plugin)
 
+        for canonical_format in self._canonical_formats:
+            self._importable.update(
+                                self._reverse_transformers[canonical_format])
+            self._exportable.update(self.transformers[canonical_format])
+
+        for importable_format in self._importable:
+            if(isinstance(importable_format,
+                          qiime2.plugin.model.file_format._FileFormat)):
+                self._importable.add(self._ff_to_sfdf[importable_format])
+
+        for exportable_format in self._exportable:
+            if(isinstance(exportable_format,
+                          qiime2.plugin.model.file_format._FileFormat)):
+                self._exportable.add(self._ff_to_sfdf[exportable_format])
+
     def _integrate_plugin(self, plugin):
-        for type_name, type_record in plugin.types.items():
-            if type_name in self.semantic_types:
-                conflicting_type_record = self.semantic_types[type_name]
+        for type_name, type_record in plugin.type_fragments.items():
+            if type_name in self.type_fragments:
+                conflicting_type_record = \
+                    self.type_fragments[type_name]
                 raise ValueError("Duplicate semantic type (%r) defined in"
                                  " plugins: %r and %r"
                                  % (type_name, type_record.plugin.name,
                                     conflicting_type_record.plugin.name))
 
-            self.semantic_types[type_name] = type_record
+            self.type_fragments[type_name] = type_record
 
         for (input, output), transformer_record in plugin.transformers.items():
             if output in self.transformers[input]:
                 raise ValueError("Transformer from %r to %r already exists."
                                  % transformer_record)
             self.transformers[input][output] = transformer_record
+            self._reverse_transformers[output][input] = transformer_record
```

Comment:
The idea is good, the name could be misleading. The transformer still runs in the same direction, it is the dictionary that is reversed. Maybe `_transformers_by_output` would help contrast with `transformers` (and maybe that should be `_transformers_by_input` in all actuality).